### ISSUE TITLE
Incorrect interpretation of $name

### ISSUE THREAD

**Issue by jvoigtlaender** [ndmitchell/hlint#475]
I suspect this has to do with (support or missing support for) Template Haskell. But looking through the issues with search term "template" didn't turn up something related, so I am opening a new issue.

The issue is that I ran hlint on some (student's) code, which is very simple code without using any language extensions. The following hlint warning came up:
```
Task04.hs:22:1: Warning: Eta reduce
Found:
  circ t = translated 5 0 $solidCircle t
Why not:
  circ = translated 5 0 $solidCircle
```
The only way this makes sense to me is that hlint thought that `$solidCircle` were an identifier. But actually the expression is (an unfortunately spaced version of) `translated 5 0 $ solidCircle t`, and replacing that by `translated 5 0 $ solidCircle` is actually wrong.

**Comment by ndmitchell** [comment]
If you add `TemplateHaskell` as an extension to the top of the file does it go away? HLint tries to guess the extensions, but it doesn't always get it right.

**Comment by jvoigtlaender** [comment]
No, adding `TemplateHaskell` as an extension does not make that hint go away. Also, I would be surprised if it did, because *with* `TemplateHaskell` being active the hint does make sense, whereas *without* `TemplateHaskell` being active it does not make sense. So the problem I am seeing is that hlint thinks `TemplateHaskell` is active when it isn't, not the other way round. I would need to explicitly say the file does *not* use `TemplateHaskell`. Indeed, that's probably what I will do for now with the student submissions.

Anyway, here a minimal collection of facts on this issue:

With the following as the whole content of the input file:
```haskell
{-# LANGUAGE NoTemplateHaskell #-}

f :: Double -> Double
f x = cos $sin x
```
no hint is produced.

With instead the following as the whole content of the input file:
```haskell
f :: Double -> Double
f x = cos $sin x
```
a wrong hint is produced (replacing `f x = cos $sin x` by `f = cos $sin`).


Whereas with the following as the whole content of the input file:
```haskell
f :: Double -> Double
f x = cos $ sin x
```
again no hint is produced.

**Comment by ndmitchell** [comment]
OK, all makes sense now. HLint tries to turn on as many extensions as it can, to parse as many files as it can. In this case `TemplateHaskell` steals syntax you were using anyway. Is there a particular reason for spacing the code as you did, without the space after `$`? To my mind, that reads rather weird, and I would always do spaces both sides. It also has the risk that if someone turns `TemplateHaskell` on at the project level your code is going to break.

I modified the manual to note that `$foo` is a particular case where HLint may misparse it.

**Comment by jvoigtlaender** [comment]
Particular reason why the code was spaced as it was? Well, I wrote that it was a student‘s code. It was their second or third Haskell program ever. Submitted as an exercise solution. So the reason is simply that the code was written by an absolute Haskell beginner, without any idea that $ could be special.

**Comment by ndmitchell** [comment]
If you are using HLint on student code, with knowledge of what extensions are turned on by default (probably none) then running with `-XHaskell2010` is probably the right approach.

**Comment by jvoigtlaender** [comment]
Yeah, obviously, in retrospect. I have for now switched over to processing all the submissions with `hlint --language=NoTemplateHaskell`, instead of just `hlint`. Actually, so far all the Haskell code they submit is CodeWorld-for-Haskell code, and I think some extensions *are* enabled in CodeWorld. But I'll see and learn what other extensions are problematic if any.

Thanks, for `hlint` and in general!
